Fix ROS2 serial communication for GRO501-1

MarcOlivierFecteau opened the pull request.

Commit:
Remove Python implementation of the ROS2 serial communication package

Commit:
Migrate ROS2 serial communication package to C++

Due to critical performance issues, caused by the additional overhead of
Python, of the serial communication package, several problems were
observed during A24. This commit migrates the Python package to a C++
package (that essentially translates the Python code into C++), which
SHOULD significantly reduce the overhead for serial communication
between the Rasperry Pi and the Arduino.

The migration seeks to eliminate the frequent loss of communication, and
heratic behavior observed in A24.

NOTES:

- The two packages used for serial communication come from other
  repositories, but adding these repositories as git submodules WOULD
  make maintaining this repository more difficult, so the `.git/`
  folder of the packages are NOT tracked.

Commit:
Fix ROS serial communication between Raspberry Pi and Arduino

Though there is still some issues with the feedback from the Arduino,
these issues are considered a lower priority, and the serial
communication package works adequately for GRO501-1. The issues MAY be
addessed for GRO501-2 validation.

NOTES:

- The changes in this commit have been tested (successfully) on the
  racecar with a Raspberry Pi 5 8GB running Ubuntu24.04;
- The requirements for GRO501-1 are respected (solutions for ROS1 were
  used for the validation: confirmed working appropriately);
- This commit does NOT contain any solution for GRO501-1: the solution
  code used for validating the serial communication and GRO501-1 were
  removed, and only the baseline ROS2 and low-level controller code are
  present.

CHANGES:

- Add `arduino_sensors` node to `teleop.launch.py` for functional
  teleoperation;
- Remove unnecessary, commented code from teleoperation launch file;
- Adjust serial communication values and logic;
- Add a code block for handling GRO501-1's "cruise control" scenario;
- Rename command and feedback topics to be more explicit;
- Remove compilation warnings from `PBUtils.cpp`;
- Add more explicit comments for GRO501-1 in `slash_teleop`.

chameau5050 commented on racecar_arduino/Controller/Controller.ino:
```diff
@@ -304,7 +304,7 @@ bool msgDiscardedLength = false;
 PBUtils pbUtils(topics);
 
-const unsigned long baud_rate = 250000;
+const unsigned long baud_rate = 115200;
```
Why are we dropping the baudrate ?

MarcOlivierFecteau replied:
The baud rate was set at `115200` in `pb2roscpp`, which is "taken as granted", so I matched the value.

IanLalonde replied:
On avait des problèmes à 250000, on n'a pas trop investigué mais il semblerait que la librairie serial en cpp supporte pas 250000

chameau5050 commented on racecar_arduino/Controller/Controller.ino:
```diff
@@ -14,7 +14,7 @@
 #include <Servo.h> 
 #define USB_USBCON
-#define IMU
+// #define IMU
```
why is this commented ?

MarcOlivierFecteau replied:
There is a compiler warning when `IMU` is defined, and considering it is not used for GRO501, I _temporarily_ commented it out.

chameau5050 commented on racecar_arduino/Controller/Controller.ino:
```diff
@@ -553,7 +562,8 @@ void sensorsCallback(unsigned long dt)
     sensorsMsg.data[2] = (float)dri_ref; // set point received by arduino
-    sensorsMsg.data[3] = (float)dri_cmd; // drive set point in volts
+    //sensorsMsg.data[3] = (float)dri_cmd; // drive set point in volts
+    sensorsMsg.data[3] = (float)Serial.available(); // drive set point in volts
```
the comment are not matching

MarcOlivierFecteau replied:
Indeed, they are not. However, the first 4-5 values of `sensorsMsg.data` are _unusable_, due to unknown issues with the serial communication package. These issues are currently NOT considered critical to the functionality of the package, so they've been put in the backlog (i.e. TODO once GRO830 has been validated).

chameau5050 commented on racecar_arduino/Controller/Controller.ino:
```diff
@@ -462,6 +462,8 @@ void setup()
   steeringServo.write(pwm_zer_ser) ;
   
+  //sensorsCallback(1);
+  
```
I would just remove that if it is not used

chameau5050 commented on racecar_arduino/Controller/Controller.ino:
```diff
@@ -541,6 +543,13 @@ void cmdCallback()
   dri_ref  = cmdMsg.data[1];  // volt or m/s or m
   ctl_mode = cmdMsg.data[2];  // 1    or 2   or 3*/
 
+  
+  unsigned long dt = time_now - time_last_high;
+  sensorsCallback(dt);
+  
+  time_last_high = time_now ;
+  enc_last_high = enc_now ;
+
```
This is unrelated to the current function

MarcOlivierFecteau replied:
Indeed, it is not. It was moved from main loop by @IanLalonde during A24, and I just copied the modifications without further thought.

chameau5050 commented on racecar_arduino/Controller/Controller.ino:
```diff
@@ -522,15 +524,15 @@ void loop()
   
-
+/*
   unsigned long dt = time_now - time_last_high;
   if (dt > time_period_high ) {
     
     sensorsCallback(dt);
     
     time_last_high = time_now ;
     enc_last_high = enc_now ;
-  }
+  }*/
```
I prefer the way it was before, so that sensor data is publish at a fixed rate

IanLalonde replied:
Il était dans le callback pour éviter que le arduino envoie trop de données rapidement au pi et remplisse le buffer. Probablement que c'est moins un probleme avec le node en cpp mais en temps normal, le callback est appelé à une fréquence fixe aussi

chameau5050 commented on racecar_autopilot/racecar_autopilot/slash_controller.py:
```diff
@@ -12,11 +12,11 @@ def __init__(self):
         # Init subscribers
         self.sub_ref = self.create_subscription(Twist, 'ctl_ref', self.read_ref, 1)
-        self.sub_prop = self.create_subscription(Float32MultiArray, 'prop_sensors', self.read_arduino, 1)
+        self.sub_prop = self.create_subscription(Float32MultiArray, 'propulsion_sensors', self.read_arduino, 1)
```
we shoud keep the old mapping to minimise mismatch in documentation

chameau5050 commented on racecar_autopilot/racecar_autopilot/slash_controller.py:
```diff
@@ -12,11 +12,11 @@ def __init__(self):
 
         # Init publishers
-        self.pub_cmd = self.create_publisher(Twist, 'prop_cmd', 1)
+        self.pub_cmd = self.create_publisher(Twist, 'propulsion_cmd', 1)
```
we shoud keep the old mapping to minimise mismatch in documentation

chameau5050 commented on racecar_bringup/launch/teleop.launch.py:
```diff
@@ -1,22 +1,16 @@
+            executable="arduino_sensors",
+            name="arduino_sensors",
+            output="screen",
+            remappings=[("raw_odom", "propulsion_sensors")],
```
I don't follow this remapping, why odom would be equal to propulsion_sensors who is define as twist in the other script

MarcOlivierFecteau replied:
I copied the code from `bringup.launch.py`, and I don't know what `arduino_sensors` is doing, and how it is doing it.

IanLalonde replied:
arduino_sensors combine les données de l'encodeur et de l'IMU pour donner le feedback d'odométrie. 
1. Prop_sensors est un array et non un twist. 
2. arduino_sensors n'a pas changé alors je le toucherais pas

chameau5050 commented on racecar_serial/libserial/Makefile:
why is there a make file and a CmakeList ?

MarcOlivierFecteau replied:
`cmake` is used to build package, and `make` is used to build the Doxygen docs (I think).

edit: `make` seems to also build the package. However, this code comes from an existing repository, so I don't think we should remove files. [here](https://github.com/wjwwood/serial) is the code's original repository.

chameau5050 commented on racecar_serial/pb2roscpp/README.md:
```diff
@@ -0,0 +1 @@
+# pb2roscpp
```
this is ours code write ? maybe a bit more info would not hurt

MarcOlivierFecteau replied:
I suggest to put the documentation of `pb2roscpp` in backlog. It would be written after the _cleanup_ of the package.

chameau5050 commented on racecar_serial/pb2roscpp/src/pb2roscpp.cpp:
```diff
@@ -0,0 +1,298 @@
+    struct message { // message structure for nanopb custom functions found on stackoverflow
+        int id;
+        std::string msg;
+    };
```
I dont think this is actualy used

chameau5050 commented on racecar_serial/pb2roscpp/src/pb2roscpp.cpp:
```diff
@@ -0,0 +1,298 @@
+        }
+
+        // Define vars for stringifing the encoded buffer
+        size_t total_bytes_encoded = ostream.bytes_written;
+        char toSend[10];
+        std::string strmsg;
+        std::string begining = "<1|";
+        std::string end = ";>";
+
+        // stringify
+        for (int j = 0; j < (int)total_bytes_encoded; ++j) {
+            sprintf(toSend, "%02X", buffer[j]);
+            strmsg += toSend;
+        }
+        // strmsg+='\0';
+        // build the message format
+        begining.append(strmsg);
+        begining.append(end);
```
I am not sure to understand why we are not just writing the buffer in a string directly since anyway the buffer is a uint8_t so it is equivalent to a char also the "%02X" make me think we are sending an unserialise version of the message I also dont get the why it is not "%03X" if it is and unserialied message

chameau5050 commented on racecar_serial/pb2roscpp/src/pb2roscpp.cpp:
```diff
@@ -0,0 +1,298 @@
+
+        // send the message
+
+        // std::cout<<(char*)begining.c_str()<<std::endl;
```
I would love to see this output for a know situation

chameau5050 commented on racecar_serial/pb2roscpp/src/pb2roscpp.cpp:
```diff
@@ -0,0 +1,298 @@
+        message.data.arg = &decodedData;
+        message.data.funcs.decode = FloatArray_decode_single_number;
+
+        if (my_serialp->available() > 0) { // wait for 1000 characters to enter the serialbuffer
```
this comment is not good

chameau5050 commented on racecar_serial/pb2roscpp/src/pb2roscpp.cpp:
```diff
@@ -0,0 +1,298 @@
+
+        if (my_serialp->available() > 0) { // wait for 1000 characters to enter the serialbuffer
+
+            my_serialp->readline(msgbuffer, MAX_MSG_LEN, ">"); // readuntil the end of a message
```
This kind of unpacking is not reliable, if my msg containt the value the uint_8 '>' this will stop before the end. Normaly a state machine is used to unpack message that have header. Reading the start bytes (<1|) than reading the packet size as the data and finaly reading the stop bytes (>) then returning only the data

chameau5050 commented on racecar_serial/pb2roscpp/src/pb2roscpp.cpp:
```diff
@@ -0,0 +1,298 @@
+
+            std_msgs::msg::Float32MultiArray prop_sensors; // instantiate the published message
+
+            decodedData.numbers[4] = my_serialp->available(); // debug IAN
```
this can be remove

chameau5050 commented on racecar_serial/pb2roscpp/src/floatarray.proto:
```diff
@@ -0,0 +1,5 @@
+syntax = "proto2";
+
+message FloatArray {
+    repeated float data = 1 [packed = true];
+}
```
why are we not defining a actual msg since we are using protobuff anyway, or at least a fixed size msg

MarcOlivierFecteau replied:
That is a **VERY** interesting question, and one I've asked myself several times since I first looked at the code in A24. I haven't been able to answer it yet, and at this point I'm afraid the code will no longer work if I change it. It seems to be the equivalent of a company's server running on a Commodore 64 since '89 ;) .

chameau5050 commented on racecar_serial/pb2roscpp/src/pb2roscpp.cpp:
```diff
@@ -0,0 +1,298 @@
+
+            pb_istream_t istream = pb_istream_from_buffer(
+                bufferin,
+                (size_t)78); // seting up decoding stream,   78 is the size of the incomming message
```
I don't like magic number

MarcOlivierFecteau replied:
Then, buckle your seatbelt, because there is some _dark magic_ in this package.

chameau5050 commented on racecar_serial/pb2roscpp/src/pb2roscpp.cpp:
I am pretty sure someting is wrong with the usage of protobuff here can you verify this is the simplest implemetantion

MarcOlivierFecteau replied:
I too am quite certain there is WAY too much voodoo in this package for our purposes. I COULD check out the code, and simplify it, but I think the best course of actions for now is to concentrate our efforts on making GRO830 work, and THEN reducing the repository's code debt.

IanLalonde replied:
Ouin considérant que à la base, je m'étais inspiré du code de communication qu'on avait dans mon projet de fin de bac (que j'avais pas écrit) que j'ai adapté pour le racecar et que après Thomas-Xavier a traduit en cpp, il y a clairement des incantations non catholiques de présentes.

chameau5050 reviewed (changes requested):
Thank you for the PR ! I still got a couple interogation worth investigating more in depth before approuving

Commit:
Apply some of the suggested modifications for #18

The applied modifications were the "low-hanging fruits" for the serial
communication. The remaining modifications suggested are categorized
as "lower-priority", and are put in the backlog. Once GRO830 is
validated, the backloged issues MAY then be addressed.

CHANGES:

- `Controller.ino`:
  - Define `IMU`;
  - Remove unused variable `stat`;
  - Remove unused, commented out code;
  - Move `sensorsCallback` code block back in `loop`;
  - Fix mismatched comment in `sensorsCallback`.
- Revert to legacy topic names;
- Remove (some) unused/commented out code in `pb2roscpp`.

MarcOlivierFecteau committed:
Merge branch 'SherbyRobotics:ros2' into fix/serial-communication

MarcOlivierFecteau committed:
Merge branch 'SherbyRobotics:ros2' into fix/serial-communication
